### DISCUSSION TITLE
Budget-style buckets

### DISCUSSION THREAD

**Discussion by andyvandesand** [TheAxelander/OpenBudgeteer#176] [General]
After playing around with the app for a while I wasn't able to make it work exactly how I wanted; I believe the current Bucket options are quite different to what I was picturing. I think I got the idea of how the current setup is supposed to work, but that's not really working for me - although it's of course entirely possible that I'm using it somehow incorrectly...

What I am looking for, and have been using in other apps and spreadsheets previously, is a budget/depletion-style bucket, which gets "filled" at the beginning of each month with a certain budget and then deplets throughout the month. If it gets emptied and into negative amounts, a variation needs to be created to move money from another bucket into the negative bucket. I think the current approach in the app is a bit different.

Based on this I have created another type of bucket to allow this behaviour. If this is actually something others are interested in I'm more than happy to add the branch. Additionally I created a webservice which provides the bucket groups together with their totals, which for my setup are linked in to other systems to be able to quickly check whether there is room in the budget for purchases.

**Comment by TheAxelander** [comment]
So the way how you explained it, is at the end the way how it works.
Let's focus just on the `Standard Bucket` as the others just automate the Input, which you can also do manually.

You have a certain amount of `Budget` which comes from all Transaction which have been marked as `Income`. This budget you can put into a Bucket via the `In/Out` field, the total amount is summarized as `In`. The Bucket now has a positive `Balance`.

If an expense Transaction will be now assigned to this bucket, money will be taken out again (consumed), which you can see as `Activity`.
Depending on the amount, the bucket has now a positive, negative or 0-based balance. In case it's negative, it's highly recommended to clear it (at least to 0). On the Bucket Page this is displayed on the top as `Negative Bucket Balance` across all Buckets.

All calculations are done on a month base. So if you switch to another month you will see again different values, especially for `Want`, which is based from these "automate Buckets" and gives you a hint how much money is expected to be put into the Bucket for the respective month. Using these type of Buckets is completely optional.

To maybe explain the use case "move money to another bucket":

Beside putting money into a Bucket you can also take it out again by entering a negative amount in the `In/Out` field. The money now goes back to the `Budget` and can be now used for another Bucket. All in all you can see `Budget` as a Bucket which works in the background and is not display as such like other Buckets but only as a number on the top of the Bucket Page.

**Comment by andyvandesand** [comment]
Yep, I got that far. My problem with that is that I have no way of planning ahead, this is all retrospectively for the month that has been.Slightly change the use case to having a fixed income. I can plan ahead and make a budget, so assign a fixed value to each bucket. I should now be able to display this and provide tracking information - not just how I went last month, but right now, how much I have spent is the budget that I have assigned. I have not figured out a way of doing this with the current functionality. The closest would be with the 2nd bucket type, but as you explain in the documentation already, it is really designed for fixed bills that are exactly the same every month.

**Comment by TheAxelander** [comment]
I don't get it why you can't plan ahead. You create multiple buckets and fill them with money. This is your estimate how much consumption you will potentially have in the month (or over the year if you take a long term bucket type). A bucket can even be simply a deposit like for savings.
The transactions finally just show the reality how it went. If you overspend your bucket, you have to clear it. If the amount was less than expected then you have some leftovers in there (which even could be re-used for the upcoming month).

> but as you explain in the documentation already, it is really designed for fixed bills that are exactly the same every month. 

That is not true, it doesn't have to be. These expenses can have any kind of value. The different outcome is just the final balance of the bucket (positive or negative).

I think it's overall just a misunderstanding of the concept. I developed OpenBudgeteer primarily as a "how much money I will need in future".

Beginning of every month I import all my Bank Transactions, assigned them to my Buckets, check how good my estimates were for the previous month and optionally move money from one bucket to another until I have a clean balance across all buckets.
Next I move to the upcoming/current month and put fresh money into each bucket. 
The result is that I also know how much budget I have left. This makes my life easier to decide if I can make a higher investment this time (e.g. new iPhone) or if I have the money for an additional month-based subscription. Or if I will potentially do an overspend and need to take money out of my savings deposit.

The important point to understand is: Only then money that I have on my account, I can plan for future expenses.

**Reply by andyvandesand** [reply]
Yes I understand that. But I'm talking about updates throughout the month, to get up to date information on how you are tracking. That's not very easy to see at all.

Hey don't get me wrong, it's your product, and I really like it! I'm just suggesting there might be a couple small modifications you can make to have it support a slightly different use case. I agree with you in that it kinda sorta does it, but kinda not really either. I completely understand if you don't want to take this on board, I'm just saying for me it wasn't working the way it was.

**Reply by TheAxelander** [reply]
All good, no worries.

I'm just trying to understand your use case to see where the gap is. 
Working with the app you can also do throughout the month, then you see which Buckets had already a consumption and which not. You can also move money at every time. So I'm not really sure what's missing there.

And of course I'm open for any kind of improvement if there is a use case for it, which I maybe just didn't had in my mind...

**Reply by andyvandesand** [reply]
This is how I'm using it now -
![image](https://github.com/TheAxelander/OpenBudgeteer/assets/64352749/153e6f68-2908-4a36-83cc-2cf6ef4c3d45)
![image](https://github.com/TheAxelander/OpenBudgeteer/assets/64352749/5553ee83-61a1-462c-a246-b052bb8683ce)
So there's not been any variations applied yet, but that's what I could do retrospectively to make sure the numbers add up. But I can see that I budgeted $555, spent $470.17 and have $84.83 remaining. I get that you can do something similar as is, but I couldn't get easy access to those numbers. 

This was just my first iteration of seing if I can make it work for me, which turned out to be quite easy. I now grab those numbers and have them available in HomeAssistant like this: 
![image](https://github.com/TheAxelander/OpenBudgeteer/assets/64352749/b52537d1-fa23-443e-8fcf-8d0b7fd663f2)

That's really what I was after, which means I an look at how I'm tracking while I'm on the road through existing infrastructure.

**Reply by TheAxelander** [reply]
So from what I see, is it the progress bar that you would like to see per bucket to have a better feeling on how you are on track with remaining budget/consumption etc.?

**Reply by TheAxelander** [reply]
What is the intention of the "variation" here?

**Reply by andyvandesand** [reply]
Yes the progressbar is definitely helpful. But even without that I couldn't get the budget vs spending vs remaining on a running basis - right now I'd want to know where I'm at for September, not how I went in August.

Variation is same as it was really, the wording just makes more sense to me - if I add a BucketMovement to even out the different buckets it'd show in here, just to indicate that the balance includes this.

**Reply by TheAxelander** [reply]
Alright, so let's take a look at September and ignore August completely.

You added some money into a Bucket on Sep-01, forecasting how much money you think you will need. `In` shows this input, `Balance` shows how much is left (which is right now the same as `In`).
Then you add some Transactions from maybe the last 2-3 days, assigning them to this Bucket. Now this budget will be consumed, which is displayed as `Activity`, `Balance` get's updated and shows how much is left (remaining budget). If it starts getting negative, you should consider moving money from one Bucket to another. 
With that you see how much is left for the next days in September...

What I can understand is the confusion here that there is no "direct movement" from A to B, the money will flow out of the Bucket as `BucketMovement` to the overall Budget pool (managed by `Income`) and then you move it to another one which is tracked as a second `BucketMovement` in the database.
What I can also understand is the point that if you take out money from a Bucket it will be added to `Activity`.  So you lose some track of "spent money" as expenses and movements are put together in that figure. Something I could consider as an improvement to be developed.

The figures on Bucket group level just summarizes all numbers from the Buckets below and is just some kind of statistics. In principle all Buckets work independetly.

**Reply by andyvandesand** [reply]
Yeah alright I think I get it. 

I think the view might just not be the most intuitive. Maybe there could be some pop-up help to explain the values? Same with the "distribute budget" bucket, it kinda makes sense once you work it out, but takes a while to get there! 

I am now just using some sql queries to get the values out that I need, that works.

**Reply by TheAxelander** [reply]
Good idea adding tool tips on some elements, I will follow up on that...
